Make disjuncts in instanceof and virtual method removal stable

peterschrammel opened the pull request:
These were dependent on the class loading order, causing
goto-diff to think that there was a change in the classes.

peterschrammel committed:
Make disjuncts in virtual method removal independent of class loading

The disjuncts created by virtual method removal depended
on the class loading order, causing goto-diff to think that
there was a change in the classes.

peterschrammel committed:
Regression tests for unstable instanceof and virtual method disjuncts

thk123 commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -445,6 +445,8 @@ void remove_virtual_functionst::get_functions(
         has_prefix(
           id2string(b.symbol_expr.get_identifier()), "java::java.lang.Object"))
         return true;
+      else if(a.symbol_expr.get_identifier() == b.symbol_expr.get_identifier())
```
So if the function names are equal then sort by class names? Don't the function names include the class names though?

peterschrammel replied:
`class_id` is the name of the class, whereas `symbol_expr` is the name of the target method to call (which might be defined in a parent class). We are sorting by target method and if the same target method is used by several classes (these generate the big or expression in the dispatch condition)  then we sort by these classes.

thk123 commented on src/goto-programs/class_hierarchy.h:
```diff
@@ -24,7 +25,7 @@ Date: April 2016
 {
 public:
-  typedef std::vector<irep_idt> idst;
+  typedef std::set<irep_idt> idst;
```
How does using a set help? Because the ordering is defined by the ordering of `irep_id` but isn't this still dependent on loading (since the class names will be given an index based on when first seen)? In any case, update commit message to explain why this fixes the problem

peterschrammel committed:
Make disjuncts in instanceof removal independent of class loading

The disjuncts created by instanceof removal depended
on the class loading order, causing goto-diff to think that
there was a change in the classes.

peterschrammel commented:
> Would it make sense to build a std::set instead of a vector in the first place?

I have tried it, but the way it is used by its callers makes a vector more appropriate and avoids copying.

thk123 commented:
Pending TG pointer bump
